src/quic/node_quic_buffer.h

```diff
@@ -9,95 +9,78 @@
 
-// Default non-op done handler.
-inline void default_quic_buffer_chunk_done(int status) {}
-
-inline bool IsEmptyBuffer(const uv_buf_t& buf) {
-  return buf.len == 0 || buf.base == nullptr;
-}
-
-// A quic_buffer_chunk contains the actual buffered data
-// along with a callback, and optional V8 object that
-// should be kept alive as long as the chunk is alive.
-struct quic_buffer_chunk : public MemoryRetainer {
-  MallocedBuffer<uint8_t> data_buf;
-  uv_buf_t buf;
-  done_cb done = default_quic_buffer_chunk_done;
-  size_t offset = 0;
-  size_t roffset = 0;
-  bool done_called = false;
-  std::unique_ptr<quic_buffer_chunk> next;
-
-  quic_buffer_chunk(
-      MallocedBuffer<uint8_t>&& buf_,
-      done_cb done_)
-    : quic_buffer_chunk(uv_buf_init(reinterpret_cast<char*>(buf_.data),
-                                    buf_.size),
-                        done_) {
-    data_buf = std::move(buf_);
-  }
+// A QuicBufferChunk contains the actual buffered data
+// along with a callback to be called when the data has
+// been consumed.
+class QuicBufferChunk : public MemoryRetainer {
+ public:
+  // Default non-op done handler.
+  static inline void default_done(int status);
 
-  explicit quic_buffer_chunk(uv_buf_t buf_) : buf(buf_) {}
+  // In this variant, the QuicBufferChunk owns the underlying
+  // data storage within a MaybeStackBuffer. The data will be
+  // freed when the QuicBufferChunk is destroyed.
+  inline explicit QuicBufferChunk(size_t len);
 
-  quic_buffer_chunk(
-      uv_buf_t buf_,
-      done_cb done_)
-    : quic_buffer_chunk(buf_) {
-    done = std::move(done_);
-  }
+  // In this variant, the QuicBufferChunk only maintains a
+  // pointer to the underlying data buffer. The QuicBufferChunk
+  // does not take ownership of the buffer. The done callback
+  // is invoked to let the caller know when the chunk is no
+  // longer being used.
+  inline QuicBufferChunk(uv_buf_t buf_, done_cb done_);
 
-  ~quic_buffer_chunk() override {
-    CHECK(done_called);
-  }
+  inline ~QuicBufferChunk() override;
+  inline void Done(int status);
+  inline void MemoryInfo(MemoryTracker* tracker) const override;
 
-  void Done(int status) {
-    done_called = true;
-    std::move(done)(status);
-  }
+  uint8_t* out() { return reinterpret_cast<uint8_t*>(buf_.base); }
 
-  void MemoryInfo(MemoryTracker* tracker) const override {
-    tracker->TrackFieldWithSize("buf", buf.len);
-  }
+  SET_MEMORY_INFO_NAME(QuicBufferChunk)
+  SET_SELF_SIZE(QuicBufferChunk)
 
-  SET_MEMORY_INFO_NAME(quic_buffer_chunk)
-  SET_SELF_SIZE(quic_buffer_chunk)
+ private:
+  // In practice, the amount of data stored in data_buf_ can vary
+  // broadly from double to quadruple digits. 200 should cover the
+  // most common cases but we can adjust it up or down if necessary
+  // later. Currently, data_buf_ is only used when writing handshake
+  // data. See QuicCryptoContext::WriteHandshake
+  MaybeStackBuffer<uint8_t, 200> data_buf_;
```

Comment:
I think we had this conversation before, but can you add a `TODO` comment here to not use `MaybeStackBuffer`? `QuicBufferChunk` instances are heap-allocated anyway, so the extra data could ideally go into the same heap allocation, and if there is no extra data then using `MaybeStackBuffer` adds 200 bytes of overhead per chunk

Reply:
I'll just make the change directly in this PR :-)

Reply:
(and thanks for the reminder that we did discuss that before :-D ..)